### PR TITLE
FEM-2753 Kaltura Player Playlist support

### PR DESCRIPTION
**Pull request by giladna**

### CODE REVIEW

File: tvplayer/src/main/java/com/kaltura/tvplayer/OTTMediaOptions.java [hunk 1]
```diff
@@ -1,50 +1,64 @@
...
 
     @Override
-    public MediaEntryProvider buildMediaProvider(String serverUrl, int partnerId, String ks, String referrer) {
-        OTTMediaOptions mediaOptions = this;
-        final PhoenixMediaProvider provider = new PhoenixMediaProvider(serverUrl, partnerId, ks)
-                .setAssetId(mediaOptions.assetId).setReferrer(referrer);
+    public MediaEntryProvider buildMediaProvider(String serverUrl, int partnerId) {
+
+        final PhoenixMediaProvider provider = new PhoenixMediaProvider(serverUrl, partnerId, ottMediaAsset.getKs())
+                .setAssetId(ottMediaAsset.getAssetId());
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/OTTMediaOptions.java]: Add a constructor to PhoenixMediaProvider that takes an OTTMediaAsset.

File: tvplayer/src/main/java/com/kaltura/tvplayer/playlist/CountDownOptions.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        this.shouldDisplay = shouldDisplay;
+    }
+
+    public boolean isEventSent() {
+        return eventSent;
+    }
+
+    public void setEventSent(boolean eventSent) {
+        this.eventSent = eventSent;
+    }
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/playlist/CountDownOptions.java]: Make package-local.

File: tvplayer/src/main/java/com/kaltura/tvplayer/playlist/CountDownOptions.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    public long getTimeToShowMS() {
+        return timeToShowMS;
+    }
+
+    public long getOrigTimeToShowMS() {
+        return origTimeToShowMS;
+    }
+
+    public void setTimeToShowMS(long timeToShowMS) {
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/playlist/CountDownOptions.java]: Make package-local

File: tvplayer/src/main/java/com/kaltura/tvplayer/playlist/CountDownOptions.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        this.timeToShowMS = timeToShowMS;
+    }
+
+    public long getDurationMS() {
+        return durationMS;
+    }
+
+    public void setDurationMS(long durationMS) {
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/playlist/CountDownOptions.java]: Remove

File: tvplayer/src/main/java/com/kaltura/tvplayer/playlist/CountDownOptions.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        this.durationMS = durationMS;
+    }
+
+    public boolean shouldDisplay() {
+        return shouldDisplay;
+    }
+
+    public void setShouldDisplay(boolean shouldDisplay) {
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/playlist/CountDownOptions.java]: Remove

**Review by noamtamim — CHANGES_REQUESTED** [review]

File: tvplayer/src/main/java/com/kaltura/tvplayer/playlist/BasicMediaOptions.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package com.kaltura.tvplayer.playlist;
+
+import com.kaltura.playkit.PKMediaEntry;
+import com.kaltura.playkit.PKPlaylistMedia;
+
+public class BasicMediaOptions extends PKPlaylistMedia {
+
+    private PKMediaEntry pkMediaEntry;
+    private CountDownOptions countDownOptions;
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/playlist/BasicMediaOptions.java]: - rename to playlistCountDownOptions -- everywhere
- mark as Nullable -- everywhere

File: tvplayer/src/main/java/com/kaltura/tvplayer/playlist/PlaylistEvent.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+@SuppressWarnings({"unused", "WeakerAccess"})
+public class PlaylistEvent implements PKEvent {
+
+    public static final Class<PlaylistLoaded> playListLoaded = PlaylistLoaded.class;
+    public static final Class<PlaylistStarted> playListStarted = PlaylistStarted.class;
+    public static final Class<PlaylistEnded> playListEnded = PlaylistEnded.class;
+    public static final Class<PlaylistCountDownStart> playlistCountDownStart = PlaylistCountDownStart.class;
+    public static final Class<PlaylistCountDownEnd> playlistCountDownEnd = PlaylistCountDownEnd.class;
+    public static final Class<PlaylistLoopStateChanged> playlistLoopStateChanged = PlaylistLoopStateChanged.class;
+    public static final Class<PlaylistShuffleStateChanged> playlistShuffleStateChanged = PlaylistShuffleStateChanged.class;
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/playlist/PlaylistEvent.java]: Remove shuffle

File: tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java [hunk 8]
```diff
@@ -0,0 +1,716 @@
...
+    private boolean recoverOnError;
+
+    private List<PKPlaylistMedia> origlPlaylistEntries;
+    private Map<String, PKMediaEntry> loadedMediasMap; // map of the media id and it's PKMediaEntry (ovp/ott in entryId format basic any string that ws given by user as id)
+
+    private enum CacheMediaType {
+        Prev, Current, Next
+    }
+
+    public PKPlaylistController(KalturaPlayer kalturaPlayer, PKPlaylist playlist, PKPlaylistType playlistType) {
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java:57]: Check if this playlistType parameter is needed

File: tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java [hunk 9]
```diff
@@ -0,0 +1,716 @@
...
+        }
+        if (shuffleEnabled) {
+            if (loadedMediasMap.containsKey(getCacheMediaId(CacheMediaType.Current))) {
+                return;
+            }
+        } else {
+            if (loadedMediasMap.containsKey(getCacheMediaId(CacheMediaType.Current))) {
+                return;
+            }
+        }
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java]: Remove

File: tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java [hunk 10]
```diff
@@ -0,0 +1,716 @@
...
+    private KalturaPlayer kalturaPlayer;
+    private PKPlaylist playlist;
+    private PKPlaylistType playlistType;
+    private PlaylistOptions playlistOptions;
+    private CountDownOptions countDownOptions;
+
+    private int currentPlayingIndex = -1;
+    private boolean playlistAutoContinue = true;
+    private boolean loopEnabled;
+    private boolean shuffleEnabled;
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java]: Remove

File: tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java [hunk 11]
```diff
@@ -0,0 +1,716 @@
...
+            if (kalturaPlayer.getMessageBus() == null) {
+                return false;
+            }
+            kalturaPlayer.getMessageBus().post(new PlaylistEvent.PlaylistError
+                    (new ErrorElement(errorMessage, errorCode)));
+        }
+        return isValidIndex;
+    }
+
+    private OVPMediaOptions getNextMediaOptions(int index, OVPPlaylistOptions ovpPlaylistOptions) {
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java:306]: Add comment to explain the algorithm

File: tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java [hunk 12]
```diff
@@ -0,0 +1,716 @@
...
+    public void playNext() {
+        log.d("playNext");
+        int playlistSize = playlist.getMediaListSize();
+        if (currentPlayingIndex + 1 == playlistSize) {
+            if (loopEnabled) {
+                replay();
+                return;
+            }
+            log.d("Ignore playNext - invalid index!");
+            playPrev();
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java]: Check if this is correct

File: tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java [hunk 13]
```diff
@@ -0,0 +1,716 @@
...
+    public boolean isMediaLoaded(String mediaId) {
+        log.d("isMediaLoaded mediaId = " + mediaId);
+        if (loadedMediasMap == null) {
+            return false;
+        }
+        return loadedMediasMap.containsKey(mediaId) && loadedMediasMap.get(mediaId) != null;
+    }
+
+    @Override
+    public void loop(boolean mode) {
```
Comment (noamtamim) [inline tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java]: setLoop()

File: tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java (hunk 8 above)
Comment (giladna) [reply tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java:57]: yes it is used internally only

File: tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java (hunk 11 above)
Comment (giladna) [reply tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java:306]: done

File: tvplayer/src/main/java/com/kaltura/tvplayer/playlist/BasicMediaOptions.java (hunk 6 above)
Comment (giladna) [reply tvplayer/src/main/java/com/kaltura/tvplayer/playlist/BasicMediaOptions.java]: done

File: tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java (hunk 10 above)
Comment (giladna) [reply tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java]: done

File: tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java (hunk 13 above)
Comment (giladna) [reply tvplayer/src/main/java/com/kaltura/tvplayer/PKPlaylistController.java]: done

**Review by noamtamim — APPROVED** [review]

**Pull request merged**